lib/pbench/agent/meister/__init__.py

```diff
@@ -0,0 +1,800 @@
+                          " 'hostname -f' where tool meister is running>",
+            "tools": {
+                "tool-0": [ "--opt-0", "--opt-1", ..., "--opt-N" ],
+                "tool-1": [ "--opt-0", "--opt-1", ..., "--opt-N" ],
+                ...,
+                "tool-N": [ "--opt-0", "--opt-1", ..., "--opt-N" ]
+            }
+        }
+
+    Each action message should contain three pieces of data: the action to
+    take, either start, stop, or send, the tool group to apply that action to,
+    and the directory in which to store the data. In JSON form it will look
+    like:
+
+        {
+            "action":     "<'start'|'stop'|'send'>",
+            "group":     "<tool group name>",
+            "directory": "<directory in which to store tool data>"
+        }
+
+    If the Tool Meister is running on the same host as the pbench agent
+    controller, then the Tool Meister will write it's data directly to the
+    ${benchmark_results_dir} using the controller's host name; if the Tool
+    Meister is running remotely, then it will use a local temporary directory
+    to write it's data, and will send that data to the Tool Data Sink during
+    the "send" phase.
+    """
+
+    @staticmethod
+    def fetch_params(params):
+        """Static help method that allows the method constructing a ToolMeister
+        instance to verify the parameters before we actually construct the
+        object.
+
+        The definition of the contents of a parameter block is really
+        independent of a ToolMeister implementation, but we keep this method
+        in the ToolMeister class since it is closely related to the
+        implementation.
+        """
+        try:
+            benchmark_run_dir = params["benchmark_run_dir"]
+            channel = params["channel"]
+            controller = params["controller"]
+            group = params["group"]
+            hostname = params["hostname"]
+            tools = params["tools"]
+        except KeyError as exc:
+            raise ToolMeisterError(f"Invalid parameter block, missing key {exc}")
+        else:
+            return benchmark_run_dir, channel, controller, group, hostname, tools
+
+    def __init__(self, pbench_bin, params, redis_server, logger):
+        self.logger = logger
+        self.pbench_bin = pbench_bin
+        ret_val = self.fetch_params(params)
+        (
+            self._benchmark_run_dir,
+            self._channel,
+            self._controller,
+            self._group,
+            self._hostname,
+            self._tools,
+        ) = ret_val
+        self._running_tools = dict()
+        self._rs = redis_server
+        logger.debug("pubsub")
+        self._pubsub = self._rs.pubsub()
+        logger.debug("subscribe %s", self._channel)
+        self._pubsub.subscribe(self._channel)
+        logger.debug("listen")
+        self._chan = self._pubsub.listen()
+        logger.debug("done listening")
```

Comment:
If we are going keep these dubug logs, can we make them a little elaborative or add more info.

Reply:
straight copy of toolmeister